mlflow/prophet.py

```diff
@@ -0,0 +1,296 @@
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+    :param signature: :py:class:`ModelSignature <mlflow.models.ModelSignature>`
+                      describes model input and output :py:class:`Schema <mlflow.types.Schema>`.
+                      The model signature can be :py:func:`inferred <mlflow.models.infer_signature>`
+                      from datasets with valid model input (e.g. the training dataset with target
+                      column omitted) and valid model output (e.g. model predictions generated on
+                      the training dataset), for example:
+
+                      .. code-block:: python
+
+                        from mlflow.models.signature import infer_signature
+
+                        model = Prophet().fit(df)
+                        train = model.history
+                        predictions = model.predict(model.make_future_dataframe(30))
+                        signature = infer_signature(train, predictions)
+    :param input_example: Input example provides one or several instances of valid
+                          model input. The example can be used as a hint of what data to feed the
+                          model. The given example will be converted to a Pandas DataFrame and then
+                          serialized to json using the Pandas split-oriented format. Bytes are
+                          base64-encoded.
+    :param pip_requirements: {{ pip_requirements }}
+    :param extra_pip_requirements: {{ extra_pip_requirements }}
+    """
+
+    import prophet
+    import pystan
+    import numpy
+
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    path = os.path.abspath(path)
+    if os.path.exists(path):
+        raise MlflowException(f"Path '{path}' already exists")
+    os.makedirs(path)
+    if mlflow_model is None:
+        mlflow_model = Model()
+    if signature is not None:
+        mlflow_model.signature = signature
+    if input_example is not None:
+        _save_example(mlflow_model, input_example, path)
+
+    model_data_path = os.path.join(path, _MODEL_BINARY_FILE_NAME)
+    _save_model(pr_model, model_data_path)
+
+    model_bin_kwargs = {_MODEL_BINARY_KEY: _MODEL_BINARY_FILE_NAME}
+    pyfunc.add_to_model(
+        mlflow_model, loader_module="mlflow.prophet", env=_CONDA_ENV_FILE_NAME, **model_bin_kwargs
+    )
+    flavor_conf = {
+        _MODEL_TYPE_KEY: pr_model.__class__.__name__,
+        _SAVE_FORMAT_KEY: kwargs.get("format", "pr"),
+        **model_bin_kwargs,
+    }
+    mlflow_model.add_flavor(
+        FLAVOR_NAME,
+        numpy_version=numpy.__version__,
+        pystan_version=pystan.__version__,
+        prophet_version=prophet.__version__,
+        **flavor_conf,
+    )
+    mlflow_model.save(os.path.join(path, MLMODEL_FILE_NAME))
+
+    if conda_env is None:
+        if pip_requirements is None:
+            # cannot use inferred requirements due to prophet's build process
```

Comment:
Nice inline comment! Can we elaborate here a bit more and indicate that prophet requires installation of several specific dependencies in a specific order, and dependency inference cannot guarantee that these dependencies are ordered properly?

cc @harupy who may also have input here. Another option we have is to add inferred dependencies after the list of hardcoded dependencies, though this may not be necessary since prophet doesn't support things like custom transformers.

Reply:
Added some more notes to this comment. This was due to the exhaustive permutation testing that I did by inserting dependency after dependency to the install chain until I got it working. The only passing install conf was the one where I explicitly defined all base requirements to the process to get the build to work (NOTE: setup.py _still fails_ when the grouped pip install occurs for serving - it only succeeds after running a build retry when the remainder of the env is already installed. The install process for Prophet has direct import statements for all of those packages.